{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concatenating all Ka/Ks data into one dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter output file name: Microbat_Gerbil_stress.txt\n"
     ]
    }
   ],
   "source": [
    "#Writing pipeline output to one file\n",
    "\n",
    "path = r'/Users/alexyu/Downloads/biotools/output'\n",
    "all_files = sorted(glob.glob(path + \"/*.kaks\"))\n",
    "all_files = sorted(all_files, key=lambda name: int(name[46:-5]))\n",
    "\n",
    "li = []\n",
    "\n",
    "for filename in all_files:\n",
    "    df = pd.read_csv(filename, index_col=None, header=0, sep='\\t')\n",
    "    li.append(df)\n",
    "\n",
    "data = pd.concat(li, axis=0, ignore_index=True)\n",
    "outFile = input('Enter output file name: ')\n",
    "data.to_csv('kaks/' + outFile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading from KaKs combined file\n",
    "\n",
    "inFile = input('Enter output file name: ')\n",
    "data = pd.read_csv(inFile, sep='\\t')\n",
    "data = data.drop(data.columns[0], axis=1)\n",
    "data.index += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_data = data[(data['Ka/Ks'] > 1) & (data['P-Value(Fisher)'] < 0.1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Method</th>\n",
       "      <th>Ka</th>\n",
       "      <th>Ks</th>\n",
       "      <th>Ka/Ks</th>\n",
       "      <th>P-Value(Fisher)</th>\n",
       "      <th>Length</th>\n",
       "      <th>S-Sites</th>\n",
       "      <th>N-Sites</th>\n",
       "      <th>Fold-Sites(0:2:4)</th>\n",
       "      <th>...</th>\n",
       "      <th>N-Substitutions</th>\n",
       "      <th>Fold-S-Substitutions(0:2:4)</th>\n",
       "      <th>Fold-N-Substitutions(0:2:4)</th>\n",
       "      <th>Divergence-Time</th>\n",
       "      <th>Substitution-Rate-Ratio(rTC:rAG:rTA:rCG:rTG:rCA/rCA)</th>\n",
       "      <th>GC(1:2:3)</th>\n",
       "      <th>ML-Score</th>\n",
       "      <th>AICc</th>\n",
       "      <th>Akaike-Weight</th>\n",
       "      <th>Model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMLUP00000003943|ENSMLUG00000004295|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.009069</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>42.46520</td>\n",
       "      <td>0.003718</td>\n",
       "      <td>7590</td>\n",
       "      <td>671.4530</td>\n",
       "      <td>6918.550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>61.8586</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.008286</td>\n",
       "      <td>0.590229:3.35548:0.640208:2.07413:0.455377:1</td>\n",
       "      <td>0.644876(0.560463:0.377503:0.996663)</td>\n",
       "      <td>-8815.69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ENSMLUP00000000511|ENSMLUG00000000563|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.027168</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>13.45200</td>\n",
       "      <td>0.012293</td>\n",
       "      <td>2472</td>\n",
       "      <td>210.1920</td>\n",
       "      <td>2261.810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>59.5883</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.025029</td>\n",
       "      <td>0.398398:1.16134:0.531629:1.2906:0.373677:1</td>\n",
       "      <td>0.62682(0.477549:0.402913:1)</td>\n",
       "      <td>-3071.48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>ENSMLUP00000010556|ENSMLUG00000011582|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.020992</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>49.70660</td>\n",
       "      <td>0.066571</td>\n",
       "      <td>1878</td>\n",
       "      <td>137.6090</td>\n",
       "      <td>1740.390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>35.9428</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.019485</td>\n",
       "      <td>0.408047:0.58954:0.319799:0.562268:0.0556805:1</td>\n",
       "      <td>0.649894(0.511981:0.4377:1)</td>\n",
       "      <td>-2256.07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>ENSMLUP00000018950|ENSMLUG00000023783|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.036442</td>\n",
       "      <td>0.014146</td>\n",
       "      <td>2.57608</td>\n",
       "      <td>0.060723</td>\n",
       "      <td>1758</td>\n",
       "      <td>270.3530</td>\n",
       "      <td>1487.650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>53.2439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.033013</td>\n",
       "      <td>0.492353:1.66278:0.232679:1.64701:0.021957:1</td>\n",
       "      <td>0.69222(0.586031:0.491482:0.999148)</td>\n",
       "      <td>-2231.67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>ENSMLUP00000010816|ENSMLUG00000011872|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.025446</td>\n",
       "      <td>0.000509</td>\n",
       "      <td>50.00000</td>\n",
       "      <td>0.098017</td>\n",
       "      <td>978</td>\n",
       "      <td>99.7746</td>\n",
       "      <td>878.225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>21.9501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.022901</td>\n",
       "      <td>0.85665:0.959164:0.989342:1.4056:0.481922:1</td>\n",
       "      <td>0.652352(0.576687:0.380368:1)</td>\n",
       "      <td>-1202.67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>ENSMLUP00000009586|ENSMLUG00000010517|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.057600</td>\n",
       "      <td>0.002610</td>\n",
       "      <td>22.06810</td>\n",
       "      <td>0.060322</td>\n",
       "      <td>978</td>\n",
       "      <td>67.8070</td>\n",
       "      <td>910.193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.8318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.053787</td>\n",
       "      <td>0.896886:1.29164:0.925796:1.30621:0.342711:1</td>\n",
       "      <td>0.625256(0.513804:0.361963:1)</td>\n",
       "      <td>-1298.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>ENSMLUP00000001876|ENSMLUG00000002060|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.098943</td>\n",
       "      <td>0.041628</td>\n",
       "      <td>2.37685</td>\n",
       "      <td>0.060420</td>\n",
       "      <td>891</td>\n",
       "      <td>101.8950</td>\n",
       "      <td>789.105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>73.0324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.092388</td>\n",
       "      <td>1.64358:1.75219:1.73944:2.33529:0.966059:1</td>\n",
       "      <td>0.725028(0.62963:0.545455:1)</td>\n",
       "      <td>-1273.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>ENSMLUP00000001114|ENSMLUG00000001218|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.353089</td>\n",
       "      <td>0.180419</td>\n",
       "      <td>1.95705</td>\n",
       "      <td>0.018123</td>\n",
       "      <td>696</td>\n",
       "      <td>68.8415</td>\n",
       "      <td>627.158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>162.8650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.336010</td>\n",
       "      <td>0.164853:1.00845:0.331499:1.01662:0.0357882:1</td>\n",
       "      <td>0.635908(0.555794:0.354077:0.997854)</td>\n",
       "      <td>-1257.61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>ENSMLUP00000022761|ENSMLUG00000017124|ENSMLUT0...</td>\n",
       "      <td>MA</td>\n",
       "      <td>0.418931</td>\n",
       "      <td>0.287889</td>\n",
       "      <td>1.45518</td>\n",
       "      <td>0.013478</td>\n",
       "      <td>1017</td>\n",
       "      <td>163.3630</td>\n",
       "      <td>853.637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>253.6430</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.397881</td>\n",
       "      <td>0.51449:2.04286:0.227013:1.80651:0.813962:1</td>\n",
       "      <td>0.732547(0.688791:0.50885:1)</td>\n",
       "      <td>-1925.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Sequence Method        Ka  \\\n",
       "6    ENSMLUP00000003943|ENSMLUG00000004295|ENSMLUT0...     MA  0.009069   \n",
       "134  ENSMLUP00000000511|ENSMLUG00000000563|ENSMLUT0...     MA  0.027168   \n",
       "240  ENSMLUP00000010556|ENSMLUG00000011582|ENSMLUT0...     MA  0.020992   \n",
       "293  ENSMLUP00000018950|ENSMLUG00000023783|ENSMLUT0...     MA  0.036442   \n",
       "560  ENSMLUP00000010816|ENSMLUG00000011872|ENSMLUT0...     MA  0.025446   \n",
       "598  ENSMLUP00000009586|ENSMLUG00000010517|ENSMLUT0...     MA  0.057600   \n",
       "749  ENSMLUP00000001876|ENSMLUG00000002060|ENSMLUT0...     MA  0.098943   \n",
       "844  ENSMLUP00000001114|ENSMLUG00000001218|ENSMLUT0...     MA  0.353089   \n",
       "853  ENSMLUP00000022761|ENSMLUG00000017124|ENSMLUT0...     MA  0.418931   \n",
       "\n",
       "           Ks     Ka/Ks  P-Value(Fisher)  Length   S-Sites   N-Sites  \\\n",
       "6    0.000214  42.46520         0.003718    7590  671.4530  6918.550   \n",
       "134  0.002020  13.45200         0.012293    2472  210.1920  2261.810   \n",
       "240  0.000422  49.70660         0.066571    1878  137.6090  1740.390   \n",
       "293  0.014146   2.57608         0.060723    1758  270.3530  1487.650   \n",
       "560  0.000509  50.00000         0.098017     978   99.7746   878.225   \n",
       "598  0.002610  22.06810         0.060322     978   67.8070   910.193   \n",
       "749  0.041628   2.37685         0.060420     891  101.8950   789.105   \n",
       "844  0.180419   1.95705         0.018123     696   68.8415   627.158   \n",
       "853  0.287889   1.45518         0.013478    1017  163.3630   853.637   \n",
       "\n",
       "     Fold-Sites(0:2:4)  ...  N-Substitutions  Fold-S-Substitutions(0:2:4)  \\\n",
       "6                  NaN  ...          61.8586                          NaN   \n",
       "134                NaN  ...          59.5883                          NaN   \n",
       "240                NaN  ...          35.9428                          NaN   \n",
       "293                NaN  ...          53.2439                          NaN   \n",
       "560                NaN  ...          21.9501                          NaN   \n",
       "598                NaN  ...          49.8318                          NaN   \n",
       "749                NaN  ...          73.0324                          NaN   \n",
       "844                NaN  ...         162.8650                          NaN   \n",
       "853                NaN  ...         253.6430                          NaN   \n",
       "\n",
       "     Fold-N-Substitutions(0:2:4)  Divergence-Time  \\\n",
       "6                            NaN         0.008286   \n",
       "134                          NaN         0.025029   \n",
       "240                          NaN         0.019485   \n",
       "293                          NaN         0.033013   \n",
       "560                          NaN         0.022901   \n",
       "598                          NaN         0.053787   \n",
       "749                          NaN         0.092388   \n",
       "844                          NaN         0.336010   \n",
       "853                          NaN         0.397881   \n",
       "\n",
       "     Substitution-Rate-Ratio(rTC:rAG:rTA:rCG:rTG:rCA/rCA)  \\\n",
       "6         0.590229:3.35548:0.640208:2.07413:0.455377:1      \n",
       "134        0.398398:1.16134:0.531629:1.2906:0.373677:1      \n",
       "240     0.408047:0.58954:0.319799:0.562268:0.0556805:1      \n",
       "293       0.492353:1.66278:0.232679:1.64701:0.021957:1      \n",
       "560        0.85665:0.959164:0.989342:1.4056:0.481922:1      \n",
       "598       0.896886:1.29164:0.925796:1.30621:0.342711:1      \n",
       "749         1.64358:1.75219:1.73944:2.33529:0.966059:1      \n",
       "844      0.164853:1.00845:0.331499:1.01662:0.0357882:1      \n",
       "853        0.51449:2.04286:0.227013:1.80651:0.813962:1      \n",
       "\n",
       "                                GC(1:2:3) ML-Score AICc  Akaike-Weight  Model  \n",
       "6    0.644876(0.560463:0.377503:0.996663) -8815.69  NaN            NaN    NaN  \n",
       "134          0.62682(0.477549:0.402913:1) -3071.48  NaN            NaN    NaN  \n",
       "240           0.649894(0.511981:0.4377:1) -2256.07  NaN            NaN    NaN  \n",
       "293   0.69222(0.586031:0.491482:0.999148) -2231.67  NaN            NaN    NaN  \n",
       "560         0.652352(0.576687:0.380368:1) -1202.67  NaN            NaN    NaN  \n",
       "598         0.625256(0.513804:0.361963:1) -1298.49  NaN            NaN    NaN  \n",
       "749          0.725028(0.62963:0.545455:1) -1273.46  NaN            NaN    NaN  \n",
       "844  0.635908(0.555794:0.354077:0.997854) -1257.61  NaN            NaN    NaN  \n",
       "853          0.732547(0.688791:0.50885:1) -1925.03  NaN            NaN    NaN  \n",
       "\n",
       "[9 rows x 22 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Used for Ensembl full protein genome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Formatting significant gene data\n",
    "\n",
    "gene_df = pd.DataFrame(columns=['Protein', 'Symbol', 'Pair Number', 'Hedgehog Protein', 'Platypus Protein','Ka/Ks', 'P-Value'])\n",
    "\n",
    "for index, row in sig_data.iterrows():\n",
    "    file = open(\"/Users/alexyu/Downloads/biotools/HP_OGs/OG{0}.fa\".format(index), mode='r')\n",
    "    text = file.read()\n",
    "    \n",
    "    gene_name_start = text.find(\"gene_symbol:\")\n",
    "    if gene_name_start < 0:\n",
    "        gene_name = None\n",
    "    else:\n",
    "        gene_name = text[gene_name_start + 12 : text.find(\" description:\", gene_name_start)]\n",
    "        \n",
    "    protein_desc_start = text.find(\"description:\")\n",
    "    if protein_desc_start < 0:\n",
    "        protein_desc = None\n",
    "    else:\n",
    "        end_pos = text.find(\" [\", protein_desc_start)\n",
    "        protein_desc = text[protein_desc_start + 12 : end_pos]\n",
    "        \n",
    "    hedgehog_gene = text[1 : text.find(\".\")]\n",
    "    platypus_gene = text[text.find(\">\", text.find(\".\")) + 1 : text.find(\".\", text.find(\">\", text.find(\".\")) + 1)]\n",
    "        \n",
    "    gene_df.loc[len(gene_df)] = [protein_desc, gene_name, index, hedgehog_gene, platypus_gene, row['Ka/Ks'], row['P-Value(Fisher)']]\n",
    "    \n",
    "gene_df['Pair Number'] = gene_df['Pair Number'].astype('int64')\n",
    "gene_df.set_index('Pair Number', inplace=True)\n",
    "\n",
    "gene_df = gene_df.dropna()\n",
    "gene_df.to_csv('HP_filtered_genes.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Used for Ensembl BioMart data queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter output file name: Microbat_Gerbil_stress.txt\n"
     ]
    }
   ],
   "source": [
    "gene_df = pd.DataFrame(columns=['Protein', 'Symbol', 'Pair Number', 'Microbat Protein', 'Gerbil Protein','Ka/Ks', 'P-Value'])\n",
    "\n",
    "for index, row in sig_data.iterrows():\n",
    "    \n",
    "    file = open(\"/Users/alexyu/Downloads/workDir/Output/OrthologousGroupsFasta/OG{0}.fa\".format(index), mode='r')\n",
    "    text = file.read()\n",
    "    \n",
    "    seqs = text.split('>')\n",
    "    seq1 = seqs[1]\n",
    "    seq2 = seqs[2]\n",
    "\n",
    "    desc1 = seq1.split('|')\n",
    "    gene_name = desc1[4]\n",
    "    prot_id1 = desc1[0]\n",
    "\n",
    "    desc2 = seq2.split('|')\n",
    "    prot_id2 = desc2[0]\n",
    "\n",
    "    stop = 0\n",
    "    if len(desc1[5]) > 0:\n",
    "        stop = desc1[5].find(' [')\n",
    "    prot_desc = desc1[5][:stop]\n",
    "    \n",
    "    #print('Gene symbol: %s' % gene_name)\n",
    "    #print('Protein: %s' % prot_desc)\n",
    "    #print('Microbat protein id: %s' % prot_id1)\n",
    "    #print('Prairie vole protein id: %s' % prot_id2)\n",
    "    \n",
    "    gene_df.loc[len(gene_df)] = [prot_desc, gene_name, index, prot_id1, prot_id2, row['Ka/Ks'], row['P-Value(Fisher)']]\n",
    "\n",
    "gene_df['Pair Number'] = gene_df['Pair Number'].astype('int64')\n",
    "gene_df.set_index('Pair Number', inplace=True)\n",
    "\n",
    "gene_df = gene_df.dropna()\n",
    "outFile = input('Enter output file name: ')\n",
    "gene_df.to_csv('filteredGenes/' + outFile, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
